ts/index.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * Inject raw response, so we can know if middleware has responsed.
+ */
+function makeInjectedResponse(koaCtx, /*markHandled,*/ whenEnded) {
+  let res = koaCtx.res;
+
+  res.on('close', whenEnded).on('finish', whenEnded);
+
+  let dummyRes = Object.create(res);
+  [
+    'setHeader',
+    'writeHead',
+    'write',
+    'end'
+  ].forEach(name => {
+    dummyRes[name] = function (...args) {
+      res[name](...args);
+      // koa2.0 initial assign statusCode to 404, reset to 200
+      if (res.statusCode === 404) {
+        res.statusCode = 200;
```

Comment:
You are right, in past I didn't find proper simple way to avoid that. It's bugfix for directly res.end(...) resulting 404.